crates/metadata/src/compatibility.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018-2023 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+
+use anyhow::{
+    anyhow,
+    bail,
+    Result,
+};
+use semver::{
+    Version,
+    VersionReq,
+};
+
+/// Version of the currently executing `cargo-contract` binary.
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[derive(Debug, serde::Deserialize)]
+struct Compatibility {
+    #[serde(rename = "cargo-contract")]
+    cargo_contract_compatibility: HashMap<Version, Requirements>,
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct Requirements {
+    #[serde(rename = "ink")]
+    ink_requirements: Vec<VersionReq>,
+}
+
+/// Checks whether the contract's ink! version is compatible with the cargo-contract
+/// binary.
+pub fn check_contract_ink_compatibility(ink_version: &Version) -> Result<()> {
+    let compatibility_list = include_str!("../compatibility_list.json");
+    let compatibility: Compatibility = serde_json::from_str(compatibility_list)?;
+    let cargo_contract_version =
+        semver::Version::parse(VERSION).expect("Parsing version failed");
+    let ink_req = &compatibility
+        .cargo_contract_compatibility
+        .get(&cargo_contract_version)
+        .ok_or(anyhow!(
+            "Missing compatibility configuration for cargo-contract: {}",
+            cargo_contract_version
+        ))?
+        .ink_requirements;
+
+    // Check if the ink! version matches any of the requirement
+    if !ink_req.iter().any(|req| req.matches(ink_version)) {
+        // Get matching ink! versions
+        let ink_matches = ink_req
+            .iter()
+            .map(|req| format!("'{}'", req))
+            .collect::<Vec<_>>()
+            .join(", ");
+
+        let ink_matches_message = if !ink_matches.is_empty() {
```

Comment:
If the `ink_req` is at all empty we should throw an error, that is invalid configuration.

Reply:
done